features/lorawan/lorastack/phy/LoRaPHY.cpp

```diff
@@ -123,316 +106,1317 @@ void LoRaPHY::handle_send(uint8_t *buf, uint8_t size)
+ *    data rate for every iteration.
+ *
+ *  - MAC layer will stop when maximum number of re-trials, i.e.,
+ *    MBED_CONF_LORA_NB_TRIALS is achieved.
+ *
+ * So essentially MBED_CONF_LORA_NB_TRIALS should be a multiple of range of
+ * data rates available. For example, in EU868 band, default max. data rate is
+ * DR_5 and min. data rate is DR_0, so total data rates available are 6.
+ *
+ * Hence MBED_CONF_LORA_NB_TRIALS should be a multiple of 6. Setting,
+ * MBED_CONF_LORA_NB_TRIALS = 6 would mean that every data rate will be tried
+ * exactly once starting from the largest and finishing at the smallest.
+ *
+ * PHY layers which do not have datarates scheme similar to EU band will ofcourse
+ * override this method.
+ */
+int8_t LoRaPHY::get_alternate_DR(uint8_t nb_trials)
+{
+    int8_t datarate = 0;
+    uint8_t total_nb_datrates = (phy_params.default_max_datarate - phy_params.min_tx_datarate) + 1;
+
+    uint8_t res = nb_trials % total_nb_datrates;
+
+    if (res == 0) {
+        datarate = phy_params.min_tx_datarate;
+    } else if (res == 1) {
+        datarate = phy_params.default_max_datarate;
+    } else {
+        datarate = (phy_params.default_max_datarate - res) + 1;
+    }
+
+    return datarate;
+}
+
+void LoRaPHY::calculate_backoff(backoff_params_t* calc_backoff)
+{
+    band_t *band_table = (band_t *) phy_params.bands.table;
+    channel_params_t *channel_list = phy_params.channels.channel_list;
+
+    uint8_t band_idx = channel_list[calc_backoff->channel].band;
+    uint16_t duty_cycle = band_table[band_idx].duty_cycle;
+    uint16_t join_duty_cycle = 0;
+
+    // Reset time-off to initial value.
+    band_table[band_idx].off_time = 0;
+
+    if (calc_backoff->joined == false) {
+        // Get the join duty cycle
+        if (calc_backoff->elapsed_time < 3600000) {
+            join_duty_cycle = BACKOFF_DC_1_HOUR;
+        } else if (calc_backoff->elapsed_time < (3600000 + 36000000)) {
+            join_duty_cycle = BACKOFF_DC_10_HOURS;
+        } else {
+            join_duty_cycle = BACKOFF_DC_24_HOURS;
+        }
+
+        // Apply the most restricting duty cycle
+        duty_cycle = MAX(duty_cycle, join_duty_cycle);
+    }
+
+    // No back-off if the last frame was not a join request and when the
+    // duty cycle is not enabled
+    if (calc_backoff->dc_enabled == false &&
+        calc_backoff->last_tx_was_join_req == false) {
+        band_table[band_idx].off_time = 0;
+    } else {
+        // Apply band time-off.
+        band_table[band_idx].off_time = calc_backoff->tx_toa * duty_cycle
+                - calc_backoff->tx_toa;
+    }
+}
+
+bool LoRaPHY::set_next_channel(channel_selection_params_t* params,
+                               uint8_t* channel, lorawan_time_t* time,
+                               lorawan_time_t* aggregate_timeoff)
+{
+    uint8_t channel_count = 0;
+    uint8_t delay_tx = 0;
+
+    uint8_t *enabled_channels = new uint8_t[phy_params.max_channel_cnt];
```

Comment:
I don't we need dynamic allocation here, since the maximum channel count is 16. I think this could be just:
uint8_t enabled_channels[LORA_MAX_NB_CHANNELS];

Reply:
Not necessarily. There are 72 channels in US band and 96 in China

Reply:
Ok. I'll add just an MBED_ASSERT here for now but maybe we could rethink this to be changed as static array.